### DISCUSSION TITLE
Replacements for some BLToolkit classes or methods missing in LinqToDB

### DISCUSSION THREAD

**Discussion by alexey-leonovich** [linq2db/linq2db#4434] [Q&A] [answered]
I'm migrating from BLToolkit to LinqToDB. And I couldn't find replacements for some classes or methods missing in LinqToDB.

1. `IObjectFactory.CreateInstance` has no `InitContext context` argument anymore. How to obtain `context.MappingSchema`, `context.DataSource` and `context.SourceObject` now?
2. `IDataContext.Delete()` and `IDataContext.InsertOrReplace()` cannot accept `IEnumerable<T>` anymore. Is iterating over entities collection and using `db.InsertOrReplace(inputEntities[i])` the only option left? Isn't that a performance hit?
3. We used `dbManager.SetCommand(command, params).Prepare()` - Prepare is missing now. How to prepare command for execution?
4. In BLToolkit we used `SqlQuery` to generate Insert query and `SqlQueryInfo` to change parameter names like that:
```
public static void SetCommand<TEntity>([NotNull] this DbManager dbManager, TEntity entity, string actionName)
{
    var queryInfo = new SqlQuery<TEntity>(dbManager).GetSqlQueryInfo(dbManager, typeof(TEntity), actionName);
    var queryText = queryInfo.QueryText;
    var parameters = queryInfo.GetParameters(dbManager, entity);
    for (var i = 0; i < parameters.Length; i++)
    {
        var parameterName = parameters[i].ParameterName;
        //Parameters and Where-clause for primary key
        if (parameterName.EndsWith("_P") || parameterName.EndsWith("_W"))
        {
            var newName = parameterName.Substring(0, parameterName.Length - 2);
            queryText = queryText.Replace(parameterName, newName);
            parameters[i].ParameterName = newName;
        }
    }
    dbManager.SetCommand(queryText, parameters);
}
```
`SqlQuery` and `SqlQueryInfo` are missing - what to use instead?

Thank you in advance.

**Comment by alexey-leonovich** [comment]
@MaceWindu @igor-tkachev need your help. Especially about missing InitContext context parameter in IObjectFactory.CreateInstance - I need to implement database objects cache [like that](https://rsdn.org/forum/prj.rfd/2875774.hot). Thank you in advance.

**Comment by igor-tkachev** [comment]
@alexey-leonovich, is this what you are looking for? 
https://github.com/linq2db/linq2db/blob/master/Tests/Linq/Tools/EntityServices/IdentityMapTests.cs

**Reply by alexey-leonovich** [reply, marked as answer]
I used EntityServiceInterceptor https://github.com/linq2db/linq2db/blob/master/Tests/Linq/Linq/EntityCreatedTests.cs. Thank you.
